### PR TITLE
Support UpdateStrategy and DeleteOption

### PR DESCRIPTION
**Pull request by clyang82**
fixed: https://github.com/openshift-online/maestro/issues/69

Introduce to configure the `delete_option` and `update_strategy` when creating or patching resource.

delete_option defines the option to delete the resource. It is optional when creating a resource. The propagationPolicy of `delete_option` can be:
- `Foreground` represents that the resource should be fourground deleted. This is a default value.
- `Orphan` represents that the resource is orphaned when deleting the resource.

update_strategy defines the strategy to update the resource. It is optional when creating a resource. The type of `update_strategy` can be:
- `ServerSideApply` means to update resource using server side apply with work-controller as the field manager. This is a default value.
- `Update` means to update resource by an update call.
- `CreateOnly` means do not update resource based on current manifest.
- `ReadOnly` means only check the existence of the resource based on the resource's metadata.

### CODE REVIEW

File: pkg/handlers/resource.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func (h resourceHandler) Patch(w http.ResponseWriter, r *http.Request) {
 		func() (interface{}, *errors.ServiceError) {
 			ctx := r.Context()
 			id := mux.Vars(r)["id"]
-			manifest, err := presenters.ConvertResourceManifest(patch.Manifest)
+			manifest, err := presenters.ConvertResourceManifest(patch.Manifest, patch.DeleteOption, patch.UpdateStrategy)
```
Comment (qiujian16) [inline pkg/handlers/resource.go:73]: I think we should support update these options.

File: pkg/api/openapi/model_resource.go [hunk 2]
```diff
@@ -406,6 +472,12 @@ func (o Resource) ToMap() (map[string]interface{}, error) {
 	if !IsNil(o.Manifest) {
 		toSerialize["manifest"] = o.Manifest
 	}
+	if !IsNil(o.DeleteOption) {
+		toSerialize["delete_option"] = o.DeleteOption
```
Comment (qiujian16) [inline pkg/api/openapi/model_resource.go:476]: delete_option or deleteOption?

File: pkg/api/resource_types.go [hunk 3]
```diff
@@ -121,19 +121,48 @@ func CloudEventToJSONMap(evt *cloudevents.Event) (datatypes.JSONMap, error) {
...
+
 	// create a cloud event with the manifest as the data
 	evt := cetypes.NewEventBuilder("maestro", cetypes.CloudEventsType{}).NewEvent()
 	eventPayload := &workpayload.Manifest{
-		Manifest: unstructured.Unstructured{Object: manifest},
-		DeleteOption: &workv1.DeleteOption{
-			PropagationPolicy: workv1.DeletePropagationPolicyTypeForeground,
-		},
+		Manifest:     unstructured.Unstructured{Object: manifest},
+		DeleteOption: delOption,
```
Comment (qiujian16) [inline pkg/api/resource_types.go:165]: @skeeey would you confirm deleteOption here works in agent side?

File: pkg/api/openapi/model_resource.go (hunk 2 above)
Comment (clyang82) [reply pkg/api/openapi/model_resource.go:476]: in restful api, it seems we should use `delete_option`. for example: `consumer_name` `create_at` etc https://github.com/openshift-online/maestro?tab=readme-ov-file#get-your-resource

File: pkg/handlers/resource.go (hunk 1 above)
Comment (clyang82) [reply pkg/handlers/resource.go:73]: Do you mean `PATCH` these options? if so, it was covered. I am working on adding test cases for that.

File: pkg/api/resource_types.go (hunk 3 above)
Comment (skeeey) [reply pkg/api/resource_types.go:165]: what kinds of cases are you concerned about?
we already have some test in our integration test https://github.com/open-cluster-management-io/ocm/blob/main/test/integration/cloudevents/deleteoption_test.go

**Review by qiujian16 — APPROVED** [review]

**Pull request merged**